chart/kubeapps/templates/_helpers.tpl

```diff
@@ -183,15 +183,22 @@ a defined apiServiceURL.
         {{- fail "At least one cluster must be defined." }}
     {{- end }}
     {{- range .Values.clusters }}
-        {{- if eq (.apiServiceURL | toString) "<nil>" }}
-            {{- if eq $kubeappsCluster "" }}
+        {{- if eq $kubeappsCluster "" }}
+            {{- if or (eq (.isKubeappsCluster | toString) "true") ( eq (.apiServiceURL | toString) "<nil>") }}
                 {{- $kubeappsCluster = .name }}
-            {{- else }}
-                {{- fail "Only one cluster can be specified without an apiServiceURL to refer to the cluster on which Kubeapps is installed." }}
+            {{- end }}
+        {{- else }}
+            {{- if or (eq (.isKubeappsCluster | toString) "true") ( eq (.apiServiceURL | toString) "<nil>") }}
+                {{- fail "Only one cluster can be configured using either 'isKubeappsCluster: true' or without an apiServiceURL to to refer to the cluster on which Kubeapps is installed. Please check the provided 'clusters' configuration." }}
             {{- end }}
         {{- end }}
     {{- end }}
-    {{- $kubeappsCluster }}
+    {{- if empty $kubeappsCluster }}
+        {{- fail "Unable to determine which cluster Kubeapps is installed in. Define 'isKubeappsCluster: true' or add a cluster without an apiServiceURL" }}
```

Comment:
Perhaps I'm missing something (open to chat, let's schedule a call later), but I think we still should handle this scenario there (in the chart, not in kubeops). Let's think about each possible scenario:

- Not setting `clusters` whatsoever:
  - Result: fallback to the default chat values, that is, use `- name: default` --> ergo, has a name.

```yaml
# clusters:
```

- Setting empy`clusters` :
  - Result: fails even in current master: `kubeapps/templates/kubeops-config.yaml:1:11: executing "kubeapps/templates/kubeops-config.yaml" at <len .Values.clusters>: error calling len: len of nil pointer` --> not allowed

```yaml
clusters:
```

- Single cluster (local) in `clusters` :
  - Result: name="kubeapps-cluster"

```yaml
clusters:
- name: kubeapps-cluster
```

- Single cluster (target), flag off, in `clusters` :
  - Result: fails, Unable to determine which cluster Kubeapps is installed in. Given that we are replacing the whole `clusters`, no local default cluster is described, and, consistently with setting an empty `clusters:`, it should fail.

```yaml
clusters:
- name: target-cluster
  apiServiceURL: url
  isKubeappsCluster: false
```

- Single cluster (target), flag on, in `clusters` :
  - Result:  name="target-cluster", it was the use case we aimed at solving in this PR, indeed.

```yaml
clusters:
- name: target-cluster
  apiServiceURL: url
  isKubeappsCluster: true
```
- Two clusters (target), flag on, in `clusters` :
  - Result: fails,  Only one cluster can be configured...

```yaml
clusters:
- name: target-cluster
  apiServiceURL: url
  isKubeappsCluster: on
- name: kubeapps-cluster
```

- Two clusters (target), flag off, in `clusters` :
  - Result:  name="target-cluster"
```yaml
clusters:
- name: target-cluster
  apiServiceURL: url
  isKubeappsCluster: on
- name: kubeapps-cluster
```

Am I missing any other possible situation? Why would we have to allow scenarios like this?

Reply:
> open to chat, let's schedule a call later

Yes, let's chat your Monday on one of the calls.

> Am I missing any other possible situation? Why would we have to allow scenarios like this (D)?

Because that's exactly the configuration required to target a single cluster that is not the cluster on which Kubeapps is installed. Ie. when we installed Kubeapps on a TKG management cluster (to which the user should *not* be able to install) we use a config exactly like D, listing only the cluster(s) available to the user (ie. none of them are the cluster on which Kubeapps is running). You can see this in the [related blog post I wrote](https://liveandletlearn.net/post/kubeapps-on-tkg-management-cluster/), where there are two clusters configured, but neither of which are the cluster on which Kubeapps is installed:

```yaml
# Configure the two workload clusters
clusters:
  - name: my-oidc-cluster
    apiServiceURL: https://DNS_OF_FIRST_WORKLOAD_CLUSTER_API_SERVER:6443
    certificateAuthorityData: CA_DATA_FOR_FIRST_WORKLOAD_CLUSTER_API_SERVER
  - name: second-oidc-cluster
    apiServiceURL: https://DNS_OF_SECOND_WORKLOAD_CLUSTER_API_SERVER:6443
    certificateAuthorityData: CA_DATA_FOR_SECOND_WORKLOAD_CLUSTER_API_SERVER
``` 

This is your (D), just with a second cluster that the user can target, AFAICT.

Reply:
Ah, I see below that you did get to this point of understanding from a later comment. Great.